Chess changes breakout for upstream

ekilmer opened the pull request:
DO NOT MERGE.

This is a draft PR only to comment and point out some separable features from this branch to be potentially submitted as standalone PRs and eventually merged into `master`

ekilmer committed:
Fix aarch64

ekilmer committed:
Use my capstone dev branch until fixes are merged into next

ekilmer committed:
Fix ARM

ekilmer committed:
Update Capstone commit to fix arm64 LD1 instruction immediates

ekilmer committed:
Ignore coverage tracking for defensive assertions and exceptions

feliam committed:
Merge branch 'master' into capstone-5-dev

feliam committed:
ENDBR64 as nop

feliam committed:
Add lacking x86 tests

feliam committed:
New x86 instrution test

ekilmer committed:
Disable write back once we hit our stop in Unicorn emulation

ekilmer committed:
Merge branch 'master' into capstone-5-dev

* master:
  Change the default to threading (#1779)
  Coveralls Take 2 (#1784)
  Capture return values in run.sh (#1776)
  Create a model for strncpy (#1770)
  Add doc, fix output bugs (#1769)

ekilmer committed:
Merge branch 'master' into capstone-5-dev

* master:
  Linux: Add stat method for FdLike (#1780)
  Use default handler for symbolic system call arguments (#1785)

Commit:
Make emulator reinitialize after write backs are disabled

Commit:
Merge branch 'master' into fix-emulate-step

Commit:
Fix linting on test_general

Extremely strange that my local black==19.10b0 doesn't catch this, but the one on GH actions does

Commit:
Add a Rust/Unicorn resumption test

Commit:
Make rusticorn binary actually check behavior

ekilmer committed:
Merge branch 'fix-emulate-step' into chess

* fix-emulate-step:
  Make rusticorn binary actually check behavior
  Add a Rust/Unicorn resumption test
  Fix linting on test_general
  Make emulator reinitialize after write backs are disabled
  Disable write back once we hit our stop in Unicorn emulation

ekilmer committed:
Merge branch 'master' into capstone-5-dev

* master:
  Change types.FunctionType=<class 'function'> (#1803)
  Fix test regressions (#1804)
  State Introspection API (#1775)
  Fix EVM account existence checks for selfdestruct and call (#1801)
  Add partial implementation of sendto syscall (#1791)
  crytic-compile: use latest release (#1795)
  Update gas metering for calls to empty accounts (#1774)
  Fix BitVec with symbolic offset and fix TranslatorSmtlib.unique thread safety (#1792)
  Fix Coveralls for external PRs (#1794)
  Convert plugin list to dict (#1781)
  Symbolic-length reads from symbolic sockets (#1786)
  Removing Thread unsafe global caching (#1788)
  Add Manticore native State-specific hooks (#1777)

ekilmer committed:
Merge branch 'capstone-5-dev' into chess

* capstone-5-dev:
  New x86 instrution test
  Add lacking x86 tests
  ENDBR64 as nop
  Ignore coverage tracking for defensive assertions and exceptions
  Update Capstone commit to fix arm64 LD1 instruction immediates
  Fix ARM
  Fix aarch64
  Use my capstone dev branch until fixes are merged into next

ekilmer committed:
Run CI on chess branch

ekilmer committed:
Support for pread64 syscall

Commit:
Delete duplicated test method

Commit:
Merge branch 'master' into fix-emulate-step

Commit:
Fix addresses and improve error handling

ekilmer committed:
Merge branch 'master' into capstone-5-dev

* master:
  native/cpu/x86: Add support for CPUID EAX=80000000h (#1811)

ekilmer committed:
Merge branch 'master' into chess

* master:
  Manticore 0.3.5 (#1808)
  Fix yices timeout argument (#1817)
  Detect default solver (#1820)
  Ignore Gas Calculations by Default (#1816)
  native/cpu/x86: Add support for CPUID EAX=80000000h (#1811)

ekilmer committed:
Fix issue with sphinx autodoc

Sphinx doesn't handle side-effects of importing

ekilmer committed:
Add last_executed_pc property to abstract CPU

Helpful for knowing the exact last executed instruction address.

ekilmer committed:
Optionally skip publishing mem read/writes in CPU

ekilmer committed:
Shallow copy AMD64RegFile but keep concrete register values

The copied RegisterFile should be read-only and used to keep track of
concrete register values at a certain instant

ekilmer committed:
Merge branch 'master' into chess

* master:
  Removed use of global solver from Native Memory (#2414)
  Support to use boolector as the SMT solver (#2410)
  Update CI and suggest to use pip3 instead of pip (#2409)
  Expressions use keyword-only arguments for init (#2395)
  Use Slots on all Expression objects (#2394)
  Allow double-adding exact same config option (#2397)
  Don't run OSX tests on PR
  Attempt to Fix solc Installation MacOS (#2392)
  Syscall specific hooks (#2389)
  TUI Support Infrastructure (#1620)
  Fix coveralls upload (#2387)
  docs: fix simple typo, straigth -> straight (#2381)
  Attempt to allow symbolic balances from the start (#1818)
  Fix state.cpu.PC member (#1825)
  Bump black and mypy (#1824)

ekilmer committed:
Remove call to pkg_resources that breaks custom installation

ekilmer committed:
Fix mypy

ekilmer committed:
Merge branch 'master' into chess

* master:
  Fix the generation of EVM tests (#2426)
  Disabled EVM events in testcases by default (#2417)
  added proper timeouts for cvc4 and boolector (#2418)

ekilmer committed:
Merge branch 'fix-emulate-step' into chess

* fix-emulate-step:
  Fix addresses and improve error handling
  Delete duplicated test method

ekilmer committed:
Merge branch 'master' into fix-emulate-step

* master: (22 commits)
  Fix the generation of EVM tests (#2426)
  Disabled EVM events in testcases by default (#2417)
  added proper timeouts for cvc4 and boolector (#2418)
  Removed use of global solver from Native Memory (#2414)
  Support to use boolector as the SMT solver (#2410)
  Update CI and suggest to use pip3 instead of pip (#2409)
  Expressions use keyword-only arguments for init (#2395)
  Use Slots on all Expression objects (#2394)
  Allow double-adding exact same config option (#2397)
  Don't run OSX tests on PR
  Attempt to Fix solc Installation MacOS (#2392)
  Syscall specific hooks (#2389)
  TUI Support Infrastructure (#1620)
  Fix coveralls upload (#2387)
  docs: fix simple typo, straigth -> straight (#2381)
  Attempt to allow symbolic balances from the start (#1818)
  Fix state.cpu.PC member (#1825)
  Bump black and mypy (#1824)
  Manticore 0.3.5 (#1808)
  Fix yices timeout argument (#1817)
  ...

ekilmer committed:
Fix Unicorn resume

* Update unicorn to latest 1.0.2

* Update capstone to latest 4.0.2

* Correctly process memory maps

* Synchronize data from manticore to unicorn upon resume

* Fix some issues with FS register and segments

* Fix some synchronization with CPUID instruction as compared to Unicorn

* X86 syscall instruction breakout for semantics

ekilmer committed:
Merge branch 'fix-emulate-step' into chess

* fix-emulate-step:
  Fix Unicorn resume

ekilmer committed:
Fix test missed during merge

ekilmer committed:
Fix more tests missed during merge

ekilmer committed:
staticmethods to get syscall info

ekilmer committed:
Rework some logging

ekilmer committed:
Better logging initialization

ekilmer committed:
Remember to set manticore property on state __enter__

ekilmer committed:
Better log message in sys_recvfrom

sschriner committed:
Chess heap tracking work (#2458)

* heap_tracking changes

* Updated arguments to account for registers and addresses + ran black

* Updated logging setLevel()

* remove _save_to_file

sschriner committed:
corrected read_arg in hook_malloc_library

sschriner committed:
Add location to alloc info and remove some debugging statements (#2463)

* Location data

* fix

* Clean out some logging statements

* more logging changes in hook_malloc_lib

ekilmer committed:
Kill Manticore if _any_ state encounters unrecoverable exception

This is useful for CHESS because we of how Manticore tasks are killed
through signals. Previously, a signal/exception is caught by the current
active state and then printed. This is an issue because if there is more
than one state, Manticore doesn't actually die when we want.

This change now kills Manticore when _any_ state encounters an
unrecoverable exception or assertion and operates as we expect upon a
death signal.

In the future, when upstreaming this branhc, we probably want to make
this opt-in to preserve previous functionality expectations.

ekilmer committed:
Merge branch 'master' into chess

* master:
  Install pinned version of truffle to fix CI (#2467)
  Use fixed owner and attacker accounts in multi_tx_analysis (#2464)
  Manticore 0.3.6 (#2456)
  Fix IntrospectionAPIPlugin Name (#2459)
  Portfolio of parallel solvers (#2420)
  Replace Quick mode with Thorough mode (#2457)
  Fix incorrect comparison for symbolic file wildcards (#2454)
  Reduce the number of calls to the SMT solver in EVM (#2411)
  Fixes to Unicorn emulation - start/stop/resume (#1796)
  Add support for multiple compilation units (#2444)
  Basic solver stats (#2415)

ekilmer commented on manticore/native/cpu/abstractcpu.py:
```diff
@@ -521,6 +530,7 @@ def __init__(self, regfile: RegisterFile, memory: Memory, **kwargs):
         self._instruction_cache: Dict[int, Any] = {}
         self._icount = 0
         self._last_pc = None
+        self._last_executed_pc = None
```
This is slightly different than `last_pc` and does indeed capture the last executed instruction address, whereas I had some issues with `last_pc` when trying to use it for that purpose.

ekilmer replied:
Opened PR https://github.com/trailofbits/manticore/pull/2475

ekilmer commented on manticore/native/cpu/abstractcpu.py:
```diff
@@ -723,7 +743,7 @@ def _raw_read(self, where: int, size: int = 1, force: bool = False) -> bytes:
         return data
 
-    def read_int(self, where, size=None, force=False):
+    def read_int(self, where, size=None, force=False, publish=True):
```
There are times where I did not want events/other hooks to fire when using multiple plugins, so I made it a parameter

ekilmer replied:
See this PR https://github.com/trailofbits/manticore/pull/2488

ekilmer commented on manticore/native/cpu/x86.py:
```diff
@@ -497,6 +497,16 @@ class AMD64RegFile(RegisterFile):
         "FPTAG",
     )
 
+    def __copy__(self):
+        """Custom shallow copy to create new dictionaries for concrete register
+        values lookups (snapshot). Should be read-only"""
+        cls = self.__class__
+        result = cls.__new__(cls)
+        result.__dict__.update(self.__dict__)
+        result._cache = self._cache.copy()
+        result._registers = self._registers.copy()
+        return result
```
I used this to take a snapshot of the register file and concrete values in a call stack frame

ekilmer replied:
See this PR https://github.com/trailofbits/manticore/pull/2489

ekilmer commented on manticore/native/heap_tracking/heap_syscalls.py:
```diff
@@ -0,0 +1,20 @@
+i386 = {
```
The `heap_tracking` directory is one feature

ekilmer commented on manticore/native/memory.py:
```diff
@@ -26,12 +26,13 @@
-    description="If True, throws a memory safety error if ANY concretization of a pointer is"
-    " out of bounds. Otherwise, forks into valid and invalid memory access states.",
-)
+if "fast_crash" not in consts:
+    consts.add(
+        "fast_crash",
+        default=False,
+        description="If True, throws a memory safety error if ANY concretization of a pointer is"
+        " out of bounds. Otherwise, forks into valid and invalid memory access states.",
+    )
```
I think this might not be needed anymore after https://github.com/trailofbits/manticore/commit/dd57db819b68dcf6e1752ed8f61800165818e5d2

ekilmer replied:
Removed with tests passing!

ekilmer commented on manticore/platforms/linux.py:
```diff
@@ -349,6 +349,9 @@ def tell(self) -> int:
     def seek(self, offset: int, whence: int = os.SEEK_SET) -> int:
         return self.file.seek(offset, whence)
 
+    def pread(self, count, offset):
+        return os.pread(self.fileno(), count, offset)
+
```
These syscalls should ideally have tests associated with them

ekilmer commented on manticore/platforms/linux.py:
```diff
@@ -3306,6 +3338,21 @@ def _interp_total_size(interp):
         last = load_segs[-1]
         return last.header.p_vaddr + last.header.p_memsz
 
+    @staticmethod
+    def implemented_syscalls() -> Iterable[str]:
+        import inspect
+
+        return (
+            x[0].split("sys_", 1)[1]
+            for x in inspect.getmembers(Linux, predicate=inspect.isfunction)
+            if x[0].startswith("sys_")
+        )
+
+    @staticmethod
+    def print_implemented_syscalls() -> None:
+        for syscall in Linux.implemented_syscalls():
+            print(syscall)
```
These are some helpful methods for providing more info about Manticore's syscall implementation state

ekilmer replied:
Opened a PR https://github.com/trailofbits/manticore/pull/2491

I didn't include the `print_implemented_syscalls` method because it's very easily implemented by the user and probably doesn't belong in the library

ekilmer reviewed (commented):
Left some comments on different aspects of this branch that should (or not) be made into separate PRs.

Will work on these throughout the coming days/week

sschriner commented on manticore/native/cpu/aarch64.py:
```diff
@@ -5302,6 +5255,9 @@ def get_arguments(self):
         for address in self.values_from(self._cpu.STACK):
             yield address
 
+    def get_return_reg(self):
+        return "X0"
+
```
addressed in #2470

ekilmer committed:
Merge branch 'master' into capstone-5-dev

* master: (35 commits)
  Track last_pc in StateDescriptors (#2471)
  Expose Result Register for Native CPU (#2470)
  Install pinned version of truffle to fix CI (#2467)
  Use fixed owner and attacker accounts in multi_tx_analysis (#2464)
  Manticore 0.3.6 (#2456)
  Fix IntrospectionAPIPlugin Name (#2459)
  Portfolio of parallel solvers (#2420)
  Replace Quick mode with Thorough mode (#2457)
  Fix incorrect comparison for symbolic file wildcards (#2454)
  Reduce the number of calls to the SMT solver in EVM (#2411)
  Fixes to Unicorn emulation - start/stop/resume (#1796)
  Add support for multiple compilation units (#2444)
  Basic solver stats (#2415)
  Fix the generation of EVM tests (#2426)
  Disabled EVM events in testcases by default (#2417)
  added proper timeouts for cvc4 and boolector (#2418)
  Removed use of global solver from Native Memory (#2414)
  Support to use boolector as the SMT solver (#2410)
  Update CI and suggest to use pip3 instead of pip (#2409)
  Expressions use keyword-only arguments for init (#2395)
  ...

ekilmer commented on manticore/core/smtlib/solver.py:
```diff
@@ -254,8 +254,16 @@ def send(self, cmd: str) -> None:
             logger.debug(">%s", cmd)
-        self._proc.stdout.flush()  # type: ignore
-        self._proc.stdin.write(f"{cmd}\n")  # type: ignore
+        try:
+            self._proc.stdout.flush()  # type: ignore
+            self._proc.stdin.write(f"{cmd}\n")  # type: ignore
+        except (BrokenPipeError, IOError) as e:
+            logger.critical(
+                f"Solver encountered an error trying to send commands: {e}.\n"
+                f"\tOutput: {self._proc.stdout}\n\n"
+                f"\tStderr: {self._proc.stderr}"
+            )
+            raise e
```
See this PR https://github.com/trailofbits/manticore/pull/2485

ekilmer commented on manticore/core/state.py:
```diff
@@ -238,6 +238,7 @@ def __enter__(self):
         new_state._input_symbols = list(self._input_symbols)
         new_state._context = copy.copy(self._context)
         new_state._id = None
+        new_state.manticore = self.manticore
```
See this PR https://github.com/trailofbits/manticore/pull/2486

ekilmer commented on manticore/core/worker.py:
```diff
@@ -188,6 +188,9 @@ def run(self, *args):
                         m._kill_state(current_state.id)
                         m._publish("did_kill_state", current_state, exc)
                         current_state = None
+                    # Kill Manticore if _any_ state encounters unrecoverable
+                    # exception/assertion
+                    m.kill()
```
See this PR https://github.com/trailofbits/manticore/pull/2487

ekilmer committed:
Merge branch 'master' into chess

* master:
  Install Python3.7 in Docker image (#2482)
  Prioritize LD_LIBRARY_PATH over system default path (#2476)
  track parent in state descriptors (#2479)
  Switch to Python 3.7 (#2472)
  Fix loading custom interpreters for ELF binaries (#2473)
  Create CITATION.cff
  Track last_pc in StateDescriptors (#2471)
  Expose Result Register for Native CPU (#2470)

ekilmer committed:
Merge branch 'master' into chess

* master:
  Bump Node from 13.x to 16.x (#2484)
  Copy/Snapshot a RegisterFile (#2489)
  Add more logging messages around solver errors (#2485)
  Add manticore reference property to States (#2486)
  Option to kill Manticore if any state fails on unrecoverable error (#2487)
  Optionally publish events for memory reads (#2488)

ekilmer committed:
Formatting

ekilmer committed:
Mypy fixes

ekilmer commented on manticore/platforms/linux_syscall_stubs.py:
```diff
@@ -1187,3 +1189,18 @@ def sys_wait4(self, upid, stat_addr, options, ru) -> int:
     def sys_waitid(self, which, upid, infop, options, ru) -> int:
         """ AUTOGENERATED UNIMPLEMENTED STUB """
         return self.simple_returns()
+
+    @staticmethod
+    def unimplemented_syscalls() -> Iterable[str]:
+        import inspect
+
+        return (
+            x[0].split("sys_", 1)[1]
+            for x in inspect.getmembers(SyscallStubs, predicate=inspect.isfunction)
+            if x[0].startswith("sys_")
+        )
+
+    @staticmethod
+    def print_unimplemented_syscalls() -> None:
+        for syscall in SyscallStubs.unimplemented_syscalls():
+            print(syscall)
```
This is moved in #2491 to use a more reliable and flexible source of system call information

ekilmer committed:
Merge branch 'master' into chess

* master:
  Add methods for getting list of (un)implemented system calls (#2491)
  Unpin Truffle Version, Use Latest crytic-compile (#2490)
  Add last_executed_pc property to abstract CPU (#2475)

ekilmer committed:
Remove unneeded sycall stubs helper function

ekilmer committed:
Remove unneeded workaround for fast_crash option in memory

Boyan-MILANOV committed:
Add SMT simplifications for bitvec subtraction

Boyan-MILANOV committed:
Replace operator SUB by built-in '-'

Boyan-MILANOV committed:
Merge branch 'dev-bitvecsub-simplifications' into chess

Boyan-MILANOV committed:
Add EXPLICIT fork policy (#2514)

Add a new fork policy called "EXPLICIT" that allows to specify arbitrary concrete values for the expression on whom we are forking.

* Add EXPLICIT fork policy allowing to provide our own concrete values

* Add feasibility checks for the EXPLICIT concretisation policy

* Add few type hints

* Don't pass kwargs to BaseException

* Only use lists in StateBase.concretize()

Boyan-MILANOV committed:
Fix `BitVecExtract` simplification for constant folding (#2524)

* Fix BitVecExtract simplification for constant folding

* Lint

* Add test for extract constant folding

* Lint

* Test: pass BitVecExtract args with keywords

Boyan-MILANOV committed:
x86 FXSAVE & FXRSTOR support (#2511)

* Add SMT simplifications for bitvec subtraction

* Add X86 support for FXSAVE and FXRSTOR

* Unicorn emulator: ignore floating point registers than aren't yet supported in unicorn

* Emulator: Also ignore MXCSR_MASK register

* Add logic to translate floating point registers values from (mantissa,exponent) to bitfield

Boyan-MILANOV committed:
Also ignore missing unicorn registers in the fallback emulator (#2531)

Boyan-MILANOV committed:
Revert "x86 FXSAVE & FXRSTOR support (#2511)"

This reverts commit 504ec9cbfdebe6c0f9fea7b21f09e29bb435def8. It seems that FXSAVE/FXSTORE on symbolic data causes or triggers some bugs in Manticore in the MATE tests, so we revert the commit until we figure out the root cause of the issue.

Boyan-MILANOV committed:
Revert "x86 FXSAVE & FXRSTOR support (#2511)"

This reverts commit 504ec9cbfdebe6c0f9fea7b21f09e29bb435def8. It seems that FXSAVE/FXSTORE on symbolic data causes or triggers some bugs in Manticore in the MATE tests, so we revert the commit until we figure out the root cause of the issue.

lordidiot committed:
add ENDBR32 as nop (#2532)

ekilmer committed:
Merge branch 'master' into capstone-5-dev

* master: (35 commits)
  Switch to stable Black (#2536)
  Fix typo in Manticore.linux constructor docstring (#2535)
  Revert CI changes made in #2526
  Release Manticore 0.3.7 (#2526)
  Optimise forking when there is only 1 solution (#2527)
  Epoll Implementation (#2529)
  Split off ENDBR32/64 from CHESS branch (#2533)
  Update to crytic-compile 0.2.2 (#2530)
  Also ignore missing unicorn registers in the fallback emulator (#2531)
  x86 FXSAVE & FXRSTOR support (#2511)
  Fix `BitVecExtract` simplification for constant folding (#2524)
  Add pip-audit action workflow (#2513)
  Add EXPLICIT fork policy (#2514)
  Enforce crytic-compile==0.2.1 (#2512)
  Improve namedtuple definition (#2506)
  Add SMT simplifications for bitvec subtraction (#2504)
  Fix handling of the program base address in Linux (#2500)
  Bump Sphinx version to 4.3.0 (#2503)
  Solver Improvements (#2502)
  Improves `namedtuple` definition (#2501)
  ...

ekilmer committed:
Remove duplicate x86 ENDBR64 instruction

ekilmer committed:
Fix test

ekilmer committed:
Fix more errors in tests

ekilmer committed:
Use latest tagged capstone==5.0.0rc2

ekilmer committed:
Merge branch 'master' into chess

* master:
  Switch to stable Black (#2536)
  Fix typo in Manticore.linux constructor docstring (#2535)
  Revert CI changes made in #2526
  Release Manticore 0.3.7 (#2526)
  Optimise forking when there is only 1 solution (#2527)
  Epoll Implementation (#2529)
  Split off ENDBR32/64 from CHESS branch (#2533)
  Update to crytic-compile 0.2.2 (#2530)
  Also ignore missing unicorn registers in the fallback emulator (#2531)
  x86 FXSAVE & FXRSTOR support (#2511)
  Fix `BitVecExtract` simplification for constant folding (#2524)
  Add pip-audit action workflow (#2513)
  Add EXPLICIT fork policy (#2514)
  Enforce crytic-compile==0.2.1 (#2512)
  Improve namedtuple definition (#2506)
  Add SMT simplifications for bitvec subtraction (#2504)
  Fix handling of the program base address in Linux (#2500)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trailofbits/manticore?pullRequest=2469) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/trailofbits/manticore?pullRequest=2469) before we can accept your contribution.<br/>**6** out of **7** committers have signed the CLA.<br/><br/>:white_check_mark: ekilmer<br/>:white_check_mark: feliam<br/>:white_check_mark: lordidiot<br/>:white_check_mark: kokrui<br/>:white_check_mark: Boyan-MILANOV<br/>:white_check_mark: sschriner<br/>:x: Eric Hennenfent<br/><hr/>**Eric Hennenfent** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trailofbits/manticore?pullRequest=2469) it.</sub>

ekilmer committed:
Merge branch 'capstone-5-dev' into chess

* capstone-5-dev:
  Use latest tagged capstone==5.0.0rc2
  Fix more errors in tests
  Fix test
  Remove duplicate x86 ENDBR64 instruction

ekilmer committed:
Formatting

ekilmer committed:
Remove duplicate instruction from bad merge

kokrui committed:
Fix bug in register_log_callback (#2542)

* Fix bug in register_log_callback

* Add type hints, re-arrange code

* Change CallbackStream to inherit io.StringIO

* Tests pass

lordidiot committed:
add newfstatat syscall

lordidiot committed:
add newfstatat syscall tests

ekilmer committed:
Merge branch 'add-sys-newfstatat' into test-newfstatat

ekilmer committed:
Initial implementation of sys_rseq

Still doesn't pass tests, but might be due to other issues. Pushing
this for further reference

ekilmer committed:
Fixes to CPUID to support latest glibc

We needed to add some more CPU features, as reported by CPUID
instruction to pass checks in glibc

ekilmer committed:
Hook CPUID instruction in emulator

Both the emulator and Manticore semantics for CPUID need to match
in order to correctly execute the same paths

ekilmer committed:
Use a patched version of Unicorn for CPUID hook in Python bindings

This patch is an open PR for upstream

ekilmer committed:
Update unicorn to cpuid commit

ekilmer committed:
Run CI on this branch

ekilmer committed:
Use upstream merged result of Unicorn for CPUID hook

ekilmer committed:
Merge branch 'master' into chess

* master:
  Capstone 5 updates (5.0.0rc2) (#1701)
  Add syscall implementation for `sys_newfstatat` (#2545)
  Fix newstat and newfstat (#2544)

ekilmer committed:
Merge branch 'chess' into fix-latest-glibc

* chess:
  Capstone 5 updates (5.0.0rc2) (#1701)
  Add syscall implementation for `sys_newfstatat` (#2545)

ekilmer commented on .github/workflows/ci.yml:
```diff
@@ -3,7 +3,7 @@ name: CI
   push:
     branches:
-      - master
+      - chess
```
This has been a PR. Let's revert this. We'll still run the PR tests
```suggestion
      - master
```

ekilmer committed:
Revert CI run on branch

ekilmer committed:
Fix Docker image for building unicorn

ekilmer committed:
Merge branch 'chess' into fix-latest-glibc

* chess:
  Revert CI run on branch

ekilmer committed:
Merge branch 'master' into chess

* master:
  Add `psubq` instruction (x86) (#2553)

ekilmer committed:
Merge branch 'master' into chess

* master:
  Fix plugin docs (#2554)
  Use protobuf version 3 (#2558)

ekilmer committed:
Use official Unicorn v2.0.0 release

ekilmer committed:
Merge branch 'master' into chess

* master:
  Update to Unicorn v2.0.0 (#2564)
  Retry coveralls upload (#2565)
  Use python3.7 instead of python3 (#2562)
  workflows/pip-audit: use our official action (#2561)

ekilmer committed:
Merge branch 'master' into chess

* master:
  Loosen test dependency version constraints (#2568)
  Loosen crytic-compile version restrictions (#2567)

Boyan-MILANOV committed:
Add boolean simplifications (#2563)

* Add boolean simplifications

* Fix typo

* Lint tests

* Test remove simplifications

* Comment simplification tests to check eth bench tests

* Re-enable BoolEqual simplifications

* Fix BoolNot

* Fix tests

* Fix more tests

* Re-enable other boolean simplifications and tests

* Fix tests

* Fix tests

* Add multiplication simplification

* Fix tests

* Lint

* Improve docstrings

Co-authored-by: Eric Kilmer <eric.d.kilmer@gmail.com>

Co-authored-by: Eric Kilmer <eric.d.kilmer@gmail.com>
